Make CLI autocompletion snappy

mih opened the pull request:
It basically inspects the incomplete argument list now, and
based on that can make more clever decisions what to load.

There are more possibilities for making things faster. For
example, we could load the extensions one by one, while searching
for unknown commands. But already now it feels great.

This change should have no impact on the full help generation. But it
breaks the interal operations needed to build manpages and docs.
This needs fixing, but was always ugly anyways.

Closes #6407 

TODO:

- [x] subcommand `-h` no longer displays choices (works for `--help-np`). For example, compare: `datalad run -h`
- [x] support auto-completion of parameter value choices

### Changelog
#### 💫 Enhancements and new features
- The CLI help generation has been sped up, and now also supports the completion of parameter values for a fixed set of choices.

yarikoptic commented:
Seems to not work for completing top level commands, pressing TAB after `datalad ` completes into options (and inserts `-`):
```
$> datalad -
-c               --dbg            --help           -l               --output-format  --version      
-C               -f               --help-np        --log-level      --report-status                 
--cmd            -h               --idbg           --on-failure     --report-type 
```

<details>
<summary>on maint:</summary> 

```shell
(git)lena:~datalad/datalad-maint[maint]git
$> datalad
add-archive-content    create                 -h                     rerun                
add-readme             create-sibling         --help                 run                  
addurls                create-sibling-github  --help-np              run-procedure        
aggregate-metadata     create-sibling-gitlab  --idbg                 save                 
annotate-paths         create-sibling-ria     install                search               
-c                     create-test-dataset    -l                     shell-completion     
-C                     --dbg                  --log-level            siblings             
check-dates            diff                   metadata               sshrun               
clean                  download-url           no-annex               status               
clone                  drop                   --on-failure           subdatasets          
--cmd                  export-archive         --output-format        test                 
containers-add         export-archive-ora     --pbs-runner           uninstall            
containers-list        export-to-figshare     push                   unlock               
containers-remove      extract-metadata       remove                 update               
containers-run         -f                     --report-status        --version            
copy-file              get                    --report-type          wtf                  

```
</details>
and seems to work on `master` ok.

mih commented:
Indeed! Thx. I had only tested the case with at least one char specified.

mih commented:
Completion is fixed now. And I no longer use parser instances for short and long help text generation -- which allows for not having to populate them at all. During completion, all details of the parser docs generation are now aware of completion-mode.

mih commented:
![image](https://user-images.githubusercontent.com/136479/152517509-ce1067b4-6168-403b-a582-fe618a604108.png)

yarikoptic commented:
still reviewing, but want to leave in the records, that #6407 remains unaddressed (well, PR doesn't claim to address it anyhow)
```shell
$> multitime -n 5 datalad --help-np 1>/dev/null                                                    
===> multitime results
1: datalad --help-np
            Mean        Std.Dev.    Min         Median      Max
real        1.368       0.046       1.309       1.364       1.449       
user        1.290       0.056       1.235       1.281       1.388       
sys         0.082       0.025       0.065       0.070       0.130  
```

mih commented:
It does on my system and for the benchmarks. Dont think I can do anything more

yarikoptic commented on datalad/cli/main.py:
```diff
@@ -62,6 +62,11 @@ def main(args=sys.argv):
+    if completing:
+        import shlex
+        # TODO support posix=False too?
+        args = shlex.split(os.environ.get('COMP_LINE')) or args
```
I will try to work out a test for it now.

yarikoptic replied:
pushed in two commits "varying" implementation so there is some history if someone decides to research into improving it.

yarikoptic reviewed (commented):
some comments and promises

mih commented on datalad/cli/main.py:
```diff
@@ -62,6 +62,11 @@ def main(args=sys.argv):
+    completing = "_ARGCOMPLETE" in os.environ
+    if completing:
+        import shlex
+        # TODO support posix=False too?
```
I added the todo, because I have no idea yet, and had/have no time to investigate.

mih commented on datalad/cli/helpers.py:
```diff
@@ -118,6 +62,118 @@ def __call__(self, parser, namespace, values, option_string=None):
+            '\n'.join(options),
+        )
+
+    def _get_short_help(self, parser):
```
Thx

mih commented:
@yarikoptic thx for adding the tests. Its seems one of them isn't working properly yet.

yarikoptic commented on datalad/cli/tests/test_main.py:
```diff
@@ -89,6 +92,17 @@ def run_main(args, exit_code=0, expect_stderr=False):
     return stdout, stderr
 
 
+def get_all_commands() -> list[str]:
```
```suggestion
def get_all_commands():
```

yarikoptic commented:
If I got the problematic test right (please cut paste which one next time if you saw it ) committed the fix, let's see

yarikoptic commented:
uff, there was power outage so datasets.datalad.org was down and that lead to avalanche of fails on travis.  I will restart that run -- site is back online.  appveyor just ran out of 1h window I guess since tests completed and all is good but job was announced "red".  not sure what to do about that one "reliably" so will just restart as well for now.

mih commented:
Thx! Only remaining test failure is the recent "slow" test, which is unrelated and got addressed elsewhere.

yarikoptic commented:
Yeap. And there is one Todo left in the original description, right?

mih commented:
I fixed the remaining issue, and added support for completing argument values. Now I can finally auto-complete `datalad wtf -S [TAB]` and get half of my brain's memory capacity back!

mih commented:
> Yeap. And there is one Todo left in the original description, right?

Nope, gone now.

mih commented:
Final benchmark results against `master`

![image](https://user-images.githubusercontent.com/136479/153003579-5485e384-0613-4060-9758-6e01c3f0c48d.png)

yarikoptic commented:
Coolio. Let's proceed then! Completion for everyone!!! ;)
